### PR TITLE
SLING-12315:  avoid new registrations during deactivation

### PR DESCRIPTION
**Pull request by ecolombo**
- use bundleContext as cancel flag during deactivation
- prevent new registration of resources during deactivation (necessary)
- prevent processing of new onChange events during deactivation (improvement)
- prevent scheduling of new bundle reloading jobs during deactivation (improvement)
- execute clearCache after setting bundleContext to null, to prevent new registration (improvement)

### CODE REVIEW

File: src/main/java/org/apache/sling/i18n/impl/JcrResourceBundleProvider.java [hunk 1]
```diff
@@ -436,8 +439,10 @@ protected void deactivate() {
             this.locatorPathsTracker = null;
         }
 
-        clearCache();
-        this.bundleContext = null;
+        synchronized(this) {
```
Comment (rombert) [inline src/main/java/org/apache/sling/i18n/impl/JcrResourceBundleProvider.java:442]: Why did you add this synchronized block here? The `this` lock guards the` bundleServiceRegistrations` field, which is already protected in `clearCache`.

**Review by rombert — COMMENTED** [review]
Thanks for the PR @ecolombo. Overall LGTM, but need clarification on one change.

File: src/main/java/org/apache/sling/i18n/impl/JcrResourceBundleProvider.java (hunk 1 above)
Comment (ecolombo) [reply src/main/java/org/apache/sling/i18n/impl/JcrResourceBundleProvider.java:442]: The intention was to ensure that this.bundleContext=null is executed before clearCache() and thus make sure that registerResourceBundle is inactivated before clearCache() is executed.

Otherwise, I think an ambitious compiler could let clearCache() execute in parallel or before this.bundleContext=null. In this case, a running job would keep filling bundleServiceRegistrations with references  during the long lasting for-loop in line 630 of clearCache(), and these references might not be garbage-collected.

Open for discussion ;-)

Comment (rombert) [reply src/main/java/org/apache/sling/i18n/impl/JcrResourceBundleProvider.java:442]: I think you're looking at interleaved access between `deactivate()` and `registerResourceBundle`, right? That is the only scenario where `bundleServiceRegistrations` will receive new entries.

It looks like it's theoretically possible for this scenario to happen, although I see the `bundleContext != null` checks on every code path and I am very skeptical that a runtime compiler will reorder staments in face of multi-thread volatile reads and writes.

I find the overall thread safety model of this class quite confusing ( multiple volatile fields, concurrent maps with Semaphore keys, synchronized blocks ( on `this` (!) , on a private collection ) and I think expanding the scope of the synchronized block would make things more confusing.

If you think that it's worth protecting the `bundleContext` and `bundleServiceRegistrations` together I would prefer to not make the `bundleContext` synchronized and instead guard access to those two fields all the time using the same lock. Of course, that can have performance implications so care must be taken.

Comment (jsedding) [reply src/main/java/org/apache/sling/i18n/impl/JcrResourceBundleProvider.java:442]: Agreed that the thread-safety model of the class is confusing.

That said, I consider the `bundleContext` and `bundleServiceRegistrations` fields to have a strong relation, because the registrations implicitly pertain to the bundle context. The synchronized block in deactivate makes sure that no thread-scheduling can lead to a registration being added to `bundleServiceRegistrations` once `bundleContext` has been set to `null`. Such registrations would never be unregistered, which is why I think it is important to guard against this case.

In order to simplify the thread-safety model, it might be sensible to encapsulate the concern of "bundle service registrations and deregistration" into its own class. This class would receive the bundle context in the constructor, removing the need for a `bundleContext` field in `JcrResourceBundleProvider`. All registrations, unregistrations, book keeping of registrations, and thread-safety concerns would be encapsulated. The object would be created in the activate method and closed in the deactivate method.

WDYT? I believe that could be a start towards reducing complexity.

**Comment by jsedding** [comment]
Superseeded by #15.
